apps/consulting/posts/scaling-python-banking-edition.md

```diff
@@ -0,0 +1,677 @@
+```
+
+### Manual Throttling
+
+To combat this, we decided to throttle how much work we were going to
+give the scheduler at a time. To do this we employed a thread-pool
+executor to submit the jobs. This would allow us to submit many jobs
+at once but still limit how much work was getting released to the
+scheduler. Each scheduled job would get a thread worker, and the job would hold
+a place in the pool until it was complete, after which the
+thread-pool would grab the next job and schedule it.
+
+However, we were still running into problems with overwhelming the cluster. One
+option that might work for some situations is lowering the number of thread
+workers, but this wasn't helpful for us. We had some jobs that were
+individually larger than the cluster could handle, even after chunking. This
+meant that we were still overloading the cluster when these large jobs were
+scheduled, and the rest of the time the cluster was significantly underutilized,
+making for poor efficiency and economy.
+
+So, instead of lowering the number of thread-workers in our
+thread-pool, we started checking the cluster resources manually with
+the Dask scheduler before submitting each job to make sure there was
+enough memory to accommodate it. This strategy gave us the benefit of being able to increase the
+number of thread-pool workers for better cluster utilization for smaller
+aggregations, while also allowing us to throttle the amount of
+submissions when we needed to. To implement this, we set it up so
+that when a thread-pool worker got a new task, we would:
+
+1. Estimate the required maximum resources the task requires.
+
+   - This calculation was essentially just the number of files in the task
+     multiplied by the estimated average file size.
+
+1. Activate a thread-lock so only one thread at a time could check cluster status
+   and submit a job.
+
+   - This prevents a race condition that would allow multiple thread-pool workers
+     to check cluster status at the same time and receive misleading statistics
+     about cluster utilization.
+
+1. Check cluster resources.
+1. If sufficient resources are _not_ available, **_(WHAT DO WE DO?)_**
```

Comment:
Flagging this. What happens if there aren't enough resources? I assume either the job gets returned to a queue or something; or, the worker sleeps for a period of time and then rechecks resources periodically; or, the job gets added to a separate "large job" queue/pool?

Reply:
Yeah, it's the second one. The worker sleeps for a period of time and then rechecks the resources periodically.

Reply:
I just pushed up an edit with more description on this.